src/core/common/data.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements `Data` related function
+ */
+
+#include "data.hpp"
+
+namespace ot {
+
+bool CompareBytes(const uint8_t *aFirstBuffer, const uint8_t *aSecondBuffer, uint16_t aLength, ByteMatcher aMatcher)
```

Comment:
rename to `MatchBytes`? 

We usually expect `compare` to return 0 if equal, and 1 or -1 if differ.

Reply:
We already have and use `Message::CompareByte()` which similar model.
My approach is to use term `Compare` for direct full comparison and `Match` term when there is some relaxed notion  of how the comparison is done.

Reply:
What about returning {-1, 0, 1}. This would align with `memcmp` and enable future use cases that may require sorting.

Reply:
We can extend it to do such comparison but it requires a bigger change (it won't work with the `ByteMatcher` model and we need to define a `ByteComparator` to also compare bytes). I don't think we have any immediate use-case for this so prefer to keep it simpler. I also prefer to keep this function private (for internal use by `Data` class only). I want to encourage the use `Data` and  `String` (or other types) when we need to do any such comparisons (which allows us to overload `<` and `<=`, ...  operators which makes it easier to use and the code nicer to read).

In the new initial push in this PR, I had this function itself as private and had an `extern` function prototype declaration of this in `Data`  class, but `LGTM` did not seem to like this https://github.com/openthread/openthread/pull/7189#issuecomment-974704703 (though I think technically this is valid). I thought of way around this to again make this function somewhat `private`/internal to `Data` by hiding it in common base class. I adopted this model in the latest push.

Reply:
Don't feel a strong need to return {-1, 0, 1} either. It may also increase the code size. 

Slightly feel `MatchBytes` is better since the last argument is already `ByteMatcher`. 

But both works for me.